File: src/main/java/edu/metrostate/trackingsystem/infrastructure/utils/JsonHandler.java
package edu.metrostate.trackingsystem.infrastructure.utils;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonParser;
import com.google.gson.reflect.TypeToken;
import edu.metrostate.trackingsystem.domain.models.Dealer;
import edu.metrostate.trackingsystem.domain.models.Vehicle;
import edu.metrostate.trackingsystem.infrastructure.database.IDatabaseContext;
import edu.metrostate.trackingsystem.infrastructure.logging.Logger;
import edu.metrostate.trackingsystem.infrastructure.database.DatabaseContext;
import edu.metrostate.trackingsystem.infrastructure.database.models.DealershipDatabase;

import java.io.File;
import java.io.IOException;
import java.io.FileWriter;
import java.nio.file.Files;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class JsonHandler implements IFileHandler {

    private static Logger logger = Logger.getLogger();
    private static JsonHandler instance;
    private static IDatabaseContext databaseContext;

    private JsonHandler() { }

    public static JsonHandler getInstance() {
        if (instance == null) {
            instance = new JsonHandler();
            logger = Logger.getLogger();
            databaseContext = DatabaseContext.getInstance();
        }
        return instance;
    }

    /**
     * @param file the file to be imported.
     * @return true if there was an error.
     */
    @Override
    public boolean importFile(File file) {
        try {
            var json = Files.readString(file.toPath());
            var jsonObject = JsonParser
                .parseString(json)
                .getAsJsonObject();

            var listType = new TypeToken<List<Vehicle>>() {}.getType();
            List<Vehicle> data = new Gson().fromJson(jsonObject.get("car_inventory"), listType);

            // Group by dealers
            Map<String, List<Vehicle>> groupByDealershipId = data.stream()
                    .collect(Collectors.groupingBy(Vehicle::getDealershipId));

            // Build a list of dealers
            List<Dealer> dealers = groupByDealershipId.entrySet().stream()
                    .map(e -> {
                        List<Vehicle> vehicles = e.getValue();
                        vehicles.forEach(Vehicle::normalize);
                        return new Dealer(e.getKey(), vehicles, vehicles.get(0).getDealershipName());
                    })
                    .toList();

            databaseContext.importJSON(dealers);
            logger.info("Imported successfully.");
            return true;
        } catch (Exception e) {
            logger.error(e.getMessage());
            return false;
        }
    }

    @Override
    public boolean exportFile(File file) {
        var gson = new GsonBuilder()
            .setPrettyPrinting()
            .create();
        var data = databaseContext.getVehicles();
        var wrapper = new DealershipDatabase(data);
        var json = gson.toJson(wrapper);
        var path = file.getAbsolutePath();

        if (!path.toLowerCase().endsWith(".json")) {
            file = new File(path + ".json");
        }

        try (FileWriter writer = new FileWriter(file)) {
            writer.write(json);
            logger.info("Successfully exported to " + file.getAbsolutePath());
            return true;
        } catch (IOException e) {
            logger.error("Error exporting JSON: " + e.getMessage());
            return false;
        }
    }

    /**
     * The following methods are internal processes to auto-save & load the database.json
     */
    public void saveSession() {
        Gson gson = new GsonBuilder()
            .setPrettyPrinting()
            .create();

        var data = databaseContext.getVehicles();
        var wrapper = new DealershipDatabase(data);
        var json = gson.toJson(wrapper);

        var folder = new File("src/main/resources/database");
        var file = new File(folder, "database.json");

        try (FileWriter writer = new FileWriter(file)) {
            writer.write(json);
            logger.info("Database JSON exported successfully.");
        } catch (IOException e) {
            logger.error(e.getMessage());
            System.exit(0);
        }
    }

    public boolean loadSession(File file) {
        return true;
    }
}

File: src/main/kotlin/edu/metrostate/dealership/infrastructure/utils/JsonHandler.kt
package edu.metrostate.dealership.infrastructure.utils

import com.google.gson.GsonBuilder
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.google.gson.reflect.TypeToken
import edu.metrostate.dealership.domain.models.Dealer
import edu.metrostate.dealership.domain.models.Vehicle
import edu.metrostate.dealership.infrastructure.database.Database
import edu.metrostate.dealership.infrastructure.imports.mappers.toDomainDealer
import edu.metrostate.dealership.infrastructure.imports.mappers.toDomainVehicle
import edu.metrostate.dealership.infrastructure.imports.mappers.toJsonVehicle
import edu.metrostate.dealership.infrastructure.imports.models.DatabaseWrapper
import edu.metrostate.dealership.infrastructure.imports.models.json.DealerJson
import edu.metrostate.dealership.infrastructure.imports.models.json.VehicleJson
import edu.metrostate.dealership.infrastructure.logging.Logger
import edu.metrostate.dealership.infrastructure.utils.GsonConfig.gson
import java.io.File
import java.io.FileWriter
import java.io.IOException
import java.nio.file.Files
import java.util.*


/**
 * Handles reading and writing of JSON formatted data files for the dealership management system.
 * Implements the IFileHandler interface to provide specific JSON processing capabilities.
 * This class uses the singleton pattern for application-wide access.
 */
class JsonHandler private constructor() : IFileHandler {

    /**
     * Imports dealership and vehicle data from a JSON file.
     *
     * The expected JSON format contains a "car_inventory" array of vehicle objects,
     * where each vehicle contains dealership information. The import process will:
     * 1. Parse the vehicle data
     * 2. Group vehicles by dealership
     * 3. Create dealership objects with their associated vehicles
     * 4. Import the data into the database context
     *
     * @param file The JSON file to import
     * @return true if the import was successful, false if an error occurred
     */
    override fun importFile(file: File): Boolean {
        return try {
            val json = Files.readString(file.toPath())
            val jsonObject = JsonParser.parseString(json).asJsonObject

            val listType = object : TypeToken<List<VehicleJson>>() {}.type
            val vehicles: List<VehicleJson> = gson.fromJson(jsonObject["car_inventory"], listType)

            val dealerJsonList = vehicles
                .groupBy { it.dealershipId }
                .map { (dealershipId, dealerVehicles) ->
                    DealerJson(
                        dealershipId = dealershipId,
                        name = dealerVehicles.first().dealershipName,
                        enabledForAcquisition = dealerVehicles.first().dealershipEnabled,
                        vehicles = dealerVehicles
                    )
                }

            databaseContext.importJSON(dealerJsonList)

            val dealers = dealerJsonList.map { it.toDomainDealer() }
            val domainVehicles = dealerJsonList.flatMap { it.vehicles.map { v -> v.toDomainVehicle() } }
            logger.info("Imported ${dealers.size} dealers and ${domainVehicles.size} vehicles.")
            true
        } catch (e: Exception) {
            logger.error("Failed to import file: ${e.message}")
            false
        }
    }

    /**
     * Exports the current dealership and vehicle data to a JSON file.
     *
     * The export format matches the expected import format with a "car_inventory" array
     * of vehicle objects that include dealership information. This ensures compatibility
     * with the import function.
     *
     * @param file The destination file for the export (will append .json extension if missing)
     * @return true if the export was successful, false if an error occurred
     */
    override fun exportFile(file: File): Boolean {
        var exportFile = file
        // the json structure of our database.json is not the same as the import.
        // the regular export follows the same import structure.
        val json = extractCurrentStateAsJson(isDatabase = false)
        val path = exportFile.absolutePath

        //Make sure the file has the right extension
        //Necessary because the ui does allow exporting without the proper extension
        if (!path.lowercase(Locale.getDefault()).endsWith(".json")) {
            exportFile = File("$path.json")
        }

        //Try writing the json to file
        try {
            FileWriter(exportFile).use { writer ->
                writer.write(json)
                logger.info("Successfully exported to " + exportFile.absolutePath)
                return true
            }
        } catch (e: IOException) {
            logger.error("Error exporting JSON: " + e.message)
            return false
        }
    }

    /**
     * Saves the current database state to a file named "database.json" in the application directory.
     *
     * This method is used for auto-saving functionality and persistence between application sessions.
     * The format used is the internal database format, not the import/export format.
     *
     * @throws IOException if writing to the file fails (will terminate the application)
     */
    fun saveSession() {
        val json = extractCurrentStateAsJson()

        val file = File("database.json")

        try {
            FileWriter(file).use { writer ->
                writer.write(json)
                logger.info("Database JSON exported successfully.")
            }
        } catch (e: IOException) {
            logger.error(e.message!!)
            System.exit(0)
        }
    }

    /**
     * Extracts the current state of the system as a JSON string.
     *
     * @param isDatabase Flag indicating whether to format the JSON for database storage (true)
     *                   or for export (false). Default is true.
     * @return JSON string representation of the current state
     */
    private fun extractCurrentStateAsJson(isDatabase: Boolean = true): String {
        val gson = GsonBuilder()
            .setPrettyPrinting()
            .create()

        val dealers: List<Dealer> = databaseContext.dealers
        val vehicles: List<Vehicle> = databaseContext.vehicles

        // Group vehicles by dealershipId
        val groupedVehicles = vehicles.groupBy { it.dealershipId }


        if (isDatabase) {
            // Build DealerJson list
            val dealerJsonList = dealers.map { dealer ->
                DealerJson(
                    dealershipId = dealer.dealershipId,
                    name = dealer.name,
                    enabledForAcquisition = dealer.enabledForAcquisition,
                    vehicles = groupedVehicles[dealer.dealershipId].orEmpty().map { it.toJsonVehicle() }
                )
            }
            return gson.toJson(DatabaseWrapper(dealerJsonList))
        }
        // Not isDatabase: Flatten to list of VehicleJson with dealer info
        val dealerMap = dealers.associateBy { it.dealershipId }

        val vehicleJsonList = vehicles.map { vehicle ->
            val dealer = dealerMap[vehicle.dealershipId]
            vehicle.toJsonVehicle(dealer?.enabledForAcquisition, dealer?.name)
        }

        val wrapper = JsonObject()
        wrapper.add("car_inventory", gson.toJsonTree(vehicleJsonList))
        return gson.toJson(wrapper)
    }

    /**
     * Loads a previously saved database state from a JSON file.
     *
     * This method expects the JSON file to be in the internal database format,
     * as created by the saveSession() method.
     *
     * @param file The database JSON file to load (typically "database.json")
     * @return true if the load was successful, false if an error occurred
     */
    fun loadSession(file: File): Boolean {
        return try {
            val json = Files.readString(file.toPath())
            val data = gson.fromJson(json, DatabaseWrapper::class.java)

            // Map DealerJson to Dealer
            val dealers = data.database.map { it.toDomainDealer() }

            // Flatten all VehicleJson to Vehicle (dealer ID already inside each vehicle)
            val vehicles = data.database.flatMap { it.vehicles.map { vj -> vj.toDomainVehicle() } }

            databaseContext.setDealers(dealers)
            databaseContext.setVehicles(vehicles)

            logger.info("Imported ${dealers.size} dealers and ${vehicles.size} vehicles.")
            true
        } catch (e: Exception) {
            logger.error("Failed to load DB: ${e.message}")
            false
        }
    }

    companion object {
        val instance: JsonHandler by lazy {
            JsonHandler()
        }

        val logger: Logger = Logger.logger

        private var databaseContext: Database = Database.instance
    }
}